modules/dex-oracle/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+			);
+
+			let (initial_price_0, initial_price_1) =
+				Self::get_current_price(&trading_pair).ok_or(Error::<T>::InvalidPool)?;
+			let initial_cumulative_0 = U256::zero();
+			let initial_cumulative_1 = U256::zero();
+			let now = T::Time::now();
+
+			CumulativePrices::<T>::insert(
+				&trading_pair,
+				(
+					initial_price_0,
+					initial_price_1,
+					initial_cumulative_0,
+					initial_cumulative_1,
+				),
+			);
+			Cumulatives::<T>::insert(&trading_pair, (initial_cumulative_0, initial_cumulative_1, now));
+
+			Ok(())
+		}
+
+		#[pallet::weight(10_000)]
+		#[transactional]
+		pub fn disable_cumulative(
+			origin: OriginFor<T>,
+			currency_id_a: CurrencyId,
+			currency_id_b: CurrencyId,
+		) -> DispatchResult {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			let trading_pair =
+				TradingPair::from_currency_ids(currency_id_a, currency_id_b).ok_or(Error::<T>::InvalidCurrencyId)?;
+			let _ = CumulativePrices::<T>::take(&trading_pair).ok_or(Error::<T>::CumulativePricesNotExists)?;
+			Cumulatives::<T>::remove(&trading_pair);
+
+			Ok(())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn try_update_cumulative(trading_pair: &TradingPair, pool_0: Balance, pool_1: Balance) {
+		// try updating enabled cumulative
+		if CumulativePrices::<T>::contains_key(trading_pair) {
+			Cumulatives::<T>::mutate(trading_pair, |(cumulative_0, cumulative_1, last_timestamp)| {
+				let now = T::Time::now();
+				// update cumulative only occurs once in one block
+				if *last_timestamp != now {
+					let interval: U256 = now.saturating_sub(*last_timestamp).saturated_into::<Balance>().into();
+					let pool_0_cumulative: U256 = U256::from(
+						ExchangeRate::checked_from_rational(pool_1, pool_0)
+							.unwrap_or_default()
+							.into_inner(),
+					)
+					.saturating_mul(interval);
+					let pool_1_cumulative: U256 = U256::from(
+						ExchangeRate::checked_from_rational(pool_0, pool_1)
+							.unwrap_or_default()
+							.into_inner(),
+					)
+					.saturating_mul(interval);
+
+					*cumulative_0 = cumulative_0.saturating_add(pool_0_cumulative);
+					*cumulative_1 = cumulative_1.saturating_add(pool_1_cumulative);
+					*last_timestamp = now;
+				}
+			});
+		}
+	}
+
+	fn get_current_price(trading_pair: &TradingPair) -> Option<(ExchangeRate, ExchangeRate)> {
+		let (pool_0, pool_1) = T::DEX::get_liquidity_pool(trading_pair.first(), trading_pair.second());
+		ExchangeRate::checked_from_rational(pool_1, pool_0).zip(ExchangeRate::checked_from_rational(pool_0, pool_1))
+	}
+
+	fn get_cumulative_price(trading_pair: &TradingPair) -> Option<(ExchangeRate, ExchangeRate)> {
+		Self::cumulative_prices(trading_pair).map(|(price_0, price_1, _, _)| (price_0, price_1))
+	}
+}
+
+impl<T: Config> Happened<(TradingPair, Balance, Balance)> for Pallet<T> {
+	fn happened(info: &(TradingPair, Balance, Balance)) {
+		let (trading_pair, pool_0, pool_1) = *info;
+		Self::try_update_cumulative(&trading_pair, pool_0, pool_1);
```

Comment:
why is this needed? modify liquidity shouldn't impact price?

Reply:
swap will modify the liquidity pool,  dex-module will trigger this hook when liquidity pool changed. Actually, add_liquidity/remove_liquidity doesn't need to trigger this hook because they won't impact the exchange rate

Reply:
also why do we need to update the pool on every swap? only the first one on the block is counted anyway, so we can just do it on `on_initialize`

Reply:
Actively query the pool and update cumulative in on_initialize, there will always be n * read_and_write(3, 1) and cumulative calculation, and the weight cannot be changed as fee. If handle it at hook, each `swap` will be (path_len - 1) * read_and_write(1, 0) at least, (path_len - 1) * read_and_write(2, 1) at most,  this weight can be charged as fee after re-benchmarking dex. I don't know which is better.